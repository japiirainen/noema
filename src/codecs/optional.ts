import { Codec, Input, Output } from "../Codec.js"

class OptionalCodec<C extends Codec<unknown>> extends Codec<
  Input<C> | undefined,
  Output<C> | undefined
> {
  readonly tag = "optional"
  constructor(readonly type: C) {
    super((value, ctx) =>
      value === undefined ? ctx.success(value) : type.validate(value, ctx)
    )
  }
}

export const optional = <C extends Codec<unknown>>(type: C) =>
  new OptionalCodec(type)
