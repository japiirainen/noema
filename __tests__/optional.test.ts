import { optional } from "../src/codecs/optional.js"
import { string } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

describe("optional", () => {
  test("should parse undefined or a value parsed by the underlying codec", () => {
    expectParseSuccess(optional(string), undefined)
    expectParseSuccess(optional(string), "")
  })

  test("should fail to parse other values", () => {
    expectParseFailure(optional(string), 1)
  })
})
