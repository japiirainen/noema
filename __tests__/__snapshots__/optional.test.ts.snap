// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optional should fail to parse other values 1`] = `
[
  {
    "code": "invalid_type",
    "message": "Expected value to be a string",
    "path": "",
    "value": 1,
  },
]
`;
